pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/SemaphoreAccessCoordinator.java

```diff
@@ -0,0 +1,241 @@
+  private int _maxSegmentSeqNumLoaded = -1;
+  private boolean _relyOnIdealState = false;
+  private final RealtimeTableDataManager _realtimeTableDataManager;
+  private final AtomicBoolean _isFirstTransitionProcessed;
+
+  public SemaphoreAccessCoordinator(Semaphore semaphore, boolean enforceConsumptionInOrder,
+      RealtimeTableDataManager realtimeTableDataManager) {
+    _semaphore = semaphore;
+    _lock = new ReentrantLock();
+    _condition = _lock.newCondition();
+    _enforceConsumptionInOrder = enforceConsumptionInOrder;
+    _realtimeTableDataManager = realtimeTableDataManager;
+    StreamIngestionConfig streamIngestionConfig = realtimeTableDataManager.getStreamIngestionConfig();
+    boolean trackSegmentSeqNumber = false;
+    if (streamIngestionConfig != null) {
+      trackSegmentSeqNumber = streamIngestionConfig.isTrackSegmentSeqNumber();
+    }
+    // if trackSegmentSeqNumber is false, server relies on ideal state to fetch previous segment to a segment.
+    if (!trackSegmentSeqNumber) {
+      _relyOnIdealState = true;
+    }
+    _isFirstTransitionProcessed = new AtomicBoolean(false);
+  }
+
+  public void acquire(LLCSegmentName llcSegmentName)
+      throws InterruptedException {
+
+    String segmentName = llcSegmentName.getSegmentName();
+
+    if (_enforceConsumptionInOrder) {
+      waitForPrevSegment(llcSegmentName);
+    }
+
+    long startTimeMs = System.currentTimeMillis();
+    while (!_semaphore.tryAcquire(5, TimeUnit.MINUTES)) {
+      LOGGER.warn("Failed to acquire partitionGroup consumer semaphore in: {} ms. Retrying.",
+          System.currentTimeMillis() - startTimeMs);
+
+      // if segment is marked online/offline by any instance, just return the current offline -> consuming
+      // transition.
+      if (!_enforceConsumptionInOrder && !isSegmentInProgress(segmentName)) {
+        throw new SegmentAlreadyConsumedException("segment: " + segmentName + " status must be in progress");
+      }
+    }
+  }
+
+  public void release() {
+    _semaphore.release();
+  }
+
+  public Semaphore getSemaphore() {
+    return _semaphore;
+  }
+
+  public void trackSegment(LLCSegmentName llcSegmentName) {
+    _lock.lock();
+    try {
+      if (!_relyOnIdealState) {
+        _maxSegmentSeqNumLoaded = Math.max(_maxSegmentSeqNumLoaded, llcSegmentName.getSequenceNumber());
+      }
+      // notify all helix threads waiting for their offline -> consuming segment's prev segment to be loaded
+      _condition.signalAll();
+    } finally {
+      _lock.unlock();
+    }
+  }
+
+  private void waitForPrevSegment(LLCSegmentName currSegment)
+      throws InterruptedException {
+
+    long startTimeMs = System.currentTimeMillis();
+
+    if ((_relyOnIdealState) || (!_isFirstTransitionProcessed.get())) {
+      // if _relyOnIdealState or no offline -> consuming transition has been processed, it means rely on
+      // ideal state to fetch previous segment.
+      String previousSegment = getPreviousSegment(currSegment);
```

Comment:
Instead of waiting on previous segment, I feel we can simply let the previous segment signal which transition to start next
this is because the coordinator has complete view of what all transitions are waiting and can take correct decision
it doesn't need to care about missing segments and all